crates/extensions/c8y_auth_proxy/src/body.rs

```diff
@@ -1,29 +1,76 @@
 use std::future::ready;
 
-use axum::async_trait;
+use axum::body::Body;
 use axum::extract::FromRequest;
 use futures::stream::once;
 use futures::StreamExt;
+use http_body::Frame;
+use http_body_util::BodyExt;
+use http_body_util::Full;
+use hyper::body::Buf;
 use hyper::body::Bytes;
-use hyper::Body;
 use hyper::Request;
 use hyper::StatusCode;
+use tokio::sync::mpsc;
 
 pub enum PossiblySmallBody {
     Small(Bytes),
-    Large(Body),
+    Large(SyncBody<Bytes, axum::Error>),
+}
+
+/// An adapter for [`http_body::Body`] implementations to guarantee [`Sync`]
+///
+/// Because [`reqwest::Body`] is `!Sync` and we need to convert it to
+/// [`axum::body::Body`] which requires `Sync`, we need a wrapper type.
+///
+/// The wrapper spawns an async task which sends the resulting frames
+/// to a [`mpsc::channel`]. This then implements `Sync`.
+#[pin_project::pin_project]
+pub struct SyncBody<D, E>(mpsc::Receiver<Result<Frame<D>, E>>);
+
+impl<D, E> SyncBody<D, E>
+where
+    D: Send + 'static,
+    E: Send + 'static,
+{
+    pub fn new(b: impl http_body::Body<Data = D, Error = E> + Send + 'static) -> Self {
+        let (tx, rx) = mpsc::channel(10);
+        tokio::spawn(async move {
+            tokio::pin!(b);
+            while let Some(frame) = b.frame().await {
+                // unwrap because we should abort this loop if the receiver is dropped
+                tx.send(frame).await.unwrap();
+            }
+        });
+        Self(rx)
+    }
+}
+
+impl<D: Buf, E> http_body::Body for SyncBody<D, E> {
+    type Data = D;
+    type Error = E;
+
+    fn poll_frame(
+        self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Option<Result<http_body::Frame<Self::Data>, Self::Error>>> {
+        let this = self.project();
+        this.0.poll_recv(cx)
+    }
 }
 
 impl PossiblySmallBody {
-    pub fn try_clone(self) -> (Body, Option<Bytes>) {
+    pub fn try_clone(self) -> (SyncBody<Bytes, axum::Error>, Option<Bytes>) {
         match self {
-            Self::Small(bytes) => (bytes.clone().into(), Some(bytes)),
+            Self::Small(bytes) => (
+                SyncBody::new(Full::new(bytes.clone()).map_err(|i| match i {})),
```

Comment:
I don't get how matching the `Infallible` error can help to make an `impl http_body::Body()` from a `Result`.

What am I missing?

Reply:
Because `Infallible` is an uninhabited type (i.e. the enum has no variants, so a value of the type can never exist), `match i {}` is valid and evaluates to `!`, so this can be cast to whatever implmentation is necessary (in this case `axum::Error`).

Reply:
> `match i {}` evaluates to `!`

This is what I was missing.